src/libs/EmojiUtils.ts

```diff
@@ -569,6 +569,137 @@ function getSpacersIndexes(allEmojis: EmojiPickerList): number[] {
+ * Surrogate pairs (combined emojis including flags)
+ * Modifiers (different skin tones and emoji variations)
+ * @param text
+ */
+
+// Surrogate pairs (combined emojis)
+const HIGH_SURROGATE_START = 0xd800;
+const HIGH_SURROGATE_END = 0xdbff;
+const LOW_SURROGATE_START = 0xdc00;
+const ZERO_WIDTH_JOINER = 0x200d;
+
+// Regional indicator symbols (flags)
+const REGIONAL_INDICATOR_START = 0x1f1e6; // 1st letter of a two-letter country code
+const REGIONAL_INDICATOR_END = 0x1f1ff; // Last letter of a two-letter country code
+
+// Fitzpatrick scale modifiers (skin tone)
+const FITZPATRICK_SCALE_START = 0x1f3fb; // Type 1 (represents light skin tone)
+const FITZPATRICK_SCALE_END = 0x1f3ff; // Type 6 (represents dark skin tone)
+
+// Variation selectors (specific variations in the presentation of other characters)
+const VARIATION_MODIFIER_START = 0xfe00; // Request text presentation of emoji
+const VARIATION_MODIFIER_END = 0xfe0f; // Indicate that the character should be displayed as an emoji
+
+function codePointFromSurrogatePair(pair: string) {
+    const highOffset = pair.charCodeAt(0) - HIGH_SURROGATE_START;
+    const lowOffset = pair.charCodeAt(1) - LOW_SURROGATE_START;
+    // eslint-disable-next-line no-bitwise
+    return (highOffset << 10) + lowOffset + 0x10000;
+}
+
+function isZeroWidthJoiner(text: string) {
+    return text?.charCodeAt(0) === ZERO_WIDTH_JOINER;
+}
+
+function isWithinInclusiveRange(value: number, lower: number, upper: number) {
+    return value >= lower && value <= upper;
+}
+
+function isFirstOfSurrogatePair(text: string) {
+    return isWithinInclusiveRange(text?.[0].charCodeAt(0), HIGH_SURROGATE_START, HIGH_SURROGATE_END);
+}
+
+function isRegionalIndicator(text: string) {
+    return isWithinInclusiveRange(codePointFromSurrogatePair(text), REGIONAL_INDICATOR_START, REGIONAL_INDICATOR_END);
+}
+
+function isFitzpatrickModifier(text: string) {
+    return isWithinInclusiveRange(codePointFromSurrogatePair(text), FITZPATRICK_SCALE_START, FITZPATRICK_SCALE_END);
+}
+
+function isVariationSelector(text: string) {
+    return isWithinInclusiveRange(text?.charCodeAt(0), VARIATION_MODIFIER_START, VARIATION_MODIFIER_END);
+}
+
+// Define how many code units make up the character
+function nextUnits(i: number, text: string) {
+    const current = text[i];
+    // If a value at index is not part of a surrogate pair, or it is at the end take value at i
+    if (!isFirstOfSurrogatePair(current) || i === text.length - 1) {
+        return 1;
+    }
+
+    const currentPair = current + text[i + 1];
+    const nextPair = text.substring(i + 2, i + 5);
+
+    if (isRegionalIndicator(currentPair) && isRegionalIndicator(nextPair)) {
+        return 4; // Flags (combination of 2 regional indicators)
+    }
+
+    if (isFitzpatrickModifier(nextPair)) {
+        return 4; // Skin tones
+    }
+    return 2; // Variations and non-BMP characters
+}
+
+function splitTextWithEmojis(text: string): string[] {
```

Comment:
I think we could _vastly_ simplify this code if we use a combination of regex + [this logic](https://github.com/Expensify/App/pull/15611#pullrequestreview-1322663118) @roryabraham designed in the past.

First, we need to update `CONST.REGEX.EMOJIS` to include the new `d` flag [which I've implemented in the past for this issue](https://github.com/facebook/hermes/pull/968) and is available now on Hermes engine:

```diff
diff --git a/src/CONST.ts b/src/CONST.ts
index f0139d82e6..ebe0ab7549 100755
--- a/src/CONST.ts
+++ b/src/CONST.ts
@@ -1760,8 +1760,8 @@ const CONST = {
 
         // eslint-disable-next-line max-len, no-misleading-character-class
         EMOJI: /[\p{Extended_Pictographic}\u200d\u{1f1e6}-\u{1f1ff}\u{1f3fb}-\u{1f3ff}\u{e0020}-\u{e007f}\u20E3\uFE0F]|[#*0-9]\uFE0F?\u20E3/gu,
-        // eslint-disable-next-line max-len, no-misleading-character-class
-        EMOJIS: /[\p{Extended_Pictographic}](\u200D[\p{Extended_Pictographic}]|[\u{1F3FB}-\u{1F3FF}]|[\u{E0020}-\u{E007F}]|\uFE0F|\u20E3)*|[\u{1F1E6}-\u{1F1FF}]{2}|[#*0-9]\uFE0F?\u20E3/gu,
+        // eslint-disable-next-line max-len, no-misleading-character-class, no-empty-character-class
+        EMOJIS: /[\p{Extended_Pictographic}](\u200D[\p{Extended_Pictographic}]|[\u{1F3FB}-\u{1F3FF}]|[\u{E0020}-\u{E007F}]|\uFE0F|\u20E3)*|[\u{1F1E6}-\u{1F1FF}]{2}|[#*0-9]\uFE0F?\u20E3/dgu,
         // eslint-disable-next-line max-len, no-misleading-character-class
         EMOJI_SKIN_TONES: /[\u{1f3fb}-\u{1f3ff}]/gu,
 

```

---

Then, let's implement the `splitTextWithEmojis` function which will use the regex and indices feature and is based on Rory's initial idea:

```tsx
type TextWithEmoji = {
    text: string;
    isEmoji: boolean;
};

function splitTextWithEmojis(text: string): TextWithEmoji[] {
    if (!text) {
        return [];
    }

    const splitText: TextWithEmoji[] = [];
    let regexResult: RegExpExecArray | null;
    let lastMatchIndexEnd = 0;
    do {
        regexResult = CONST.REGEX.EMOJIS.exec(text);

        if (regexResult?.indices) {
            const matchIndexStart = regexResult.indices[0][0];
            const matchIndexEnd = regexResult.indices[0][1];

            if (matchIndexStart > lastMatchIndexEnd) {
                splitText.push({
                    text: text.slice(lastMatchIndexEnd, matchIndexStart),
                    isEmoji: false,
                });
            }

            splitText.push({
                text: text.slice(matchIndexStart, matchIndexEnd),
                isEmoji: true,
            });

            lastMatchIndexEnd = matchIndexEnd;
        }
    } while (regexResult !== null);

    if (lastMatchIndexEnd < text.length) {
        splitText.push({
            text: text.slice(lastMatchIndexEnd, text.length),
            isEmoji: false,
        });
    }

    return splitText;
}
```

---

You'll notice some TS errors in the function regarding `indices` not being existent. We'll need to augment those Regex types to include this new property, so let's create the `src/types/modules/lib.es2018.regexp.d.ts` file:

```tsx
/* eslint-disable @typescript-eslint/naming-convention */
/* eslint-disable @typescript-eslint/consistent-type-definitions */

declare global {
    interface Indices extends Array<[number, number]> {
        groups?: Record<string, [number, number]>;

        /**
         * The first match. This will always be present because `null` will be returned if there are no matches.
         */
        0: [number, number];
    }

    interface RegExpMatchArray {
        indices?: Indices;
    }

    interface RegExpExecArray {
        indices?: Indices;
    }
}

// We used the export {} line to mark this file as an external module
export {};

```

---

Let's test this function now! If I pass this string `Hello world 🙂🙂🙂 ! 😶‍🌫️😶‍🌫️😶‍🌫️ test 😶‍🌫️ 😶‍🌫️ test2 👍👍🏿` which has some normal and surrogate emojis, the output will be:

```json
[
    {"isEmoji":false,"text":"Hello world "},
    {"isEmoji":true,"text":"🙂"},
    {"isEmoji":true,"text":"🙂"},
    {"isEmoji":true,"text":"🙂"},
    {"isEmoji":false,"text":" ! "},
    {"isEmoji":true,"text":"😶‍🌫️"},
    {"isEmoji":true,"text":"😶‍🌫️"},
    {"isEmoji":true,"text":"😶‍🌫️"},
    {"isEmoji":false,"text":" test "},
    {"isEmoji":true,"text":"😶‍🌫️"},
    {"isEmoji":false,"text":" "},
    {"isEmoji":true,"text":"😶‍🌫️"},
    {"isEmoji":false,"text":" test2 "},
    {"isEmoji":true,"text":"👍"},
    {"isEmoji":true,"text":"👍🏿"}
]
```

You can adjust the function to just return an array withe the separated tokens if you want, could you give it a try and see if it suits your needs in the PR?

Reply:
Update: we are still testing the regex solution

Reply:
@kbieganowski I've updated my previous solution:

1. Fixed `splitTextWithEmojis` function to always create a separate regex instance on each call, thus preventing the issue we were discussing internally.
2. Simplified solution to the TS errors regarding `indices`.

Reply:
@kbieganowski After internal discussions, we noticed that having the `g` (global) flag on `CONST.REGEX.EMOJIS` can result in hard-to-detect bugs throughout the App due to the stateful nature of the regex when it has the `g` flag. For example, repeatedly calls to `test()` with same input will sometimes return different results because the regex is being used globally.

So, I would suggest to remove the `g` flag from `CONST.REGEX.EMOJIS`:

```tsx
        // eslint-disable-next-line max-len, no-misleading-character-class, no-empty-character-class
        EMOJIS: /[\p{Extended_Pictographic}](\u200D[\p{Extended_Pictographic}]|[\u{1F3FB}-\u{1F3FF}]|[\u{E0020}-\u{E007F}]|\uFE0F|\u20E3)*|[\u{1F1E6}-\u{1F1FF}]{2}|[#*0-9]\uFE0F?\u20E3/du,
```

And then adding it when cloning the regex inside `splitTextWithEmojis`:

```tsx
    // We clone the regex and add the `g` (global) flag in order to `exec()` be stateful and work properly.
    // Adding the flag directly to CONST.ts can result in bugs throughout the App due to the stateful nature
    // of the regex when it has the `g` flag.
    const emojisRegex = new RegExp(CONST.REGEX.EMOJIS, CONST.REGEX.EMOJIS.flags.concat('g'));
```

This change will require testing in all areas of the app that uses `CONST.REGEX.EMOJIS` since we are removed the `g` flag now.

I would advice doing the same for `CONST.REGEX.EMOJI` and `CONST.REGEX.EMOJI_SKIN_TONES`, that could be done in a follow-up PR to avoid regressions on this one.